autopilot/manager.go

```diff
@@ -0,0 +1,277 @@
+package autopilot
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing"
+)
+
+// ManagerCfg houses a set of values and methods that is passed to the Manager
+// for it to properly manage its autopilot agent.
+type ManagerCfg struct {
+	// Self is the public key of the lnd instance. It is used to making
+	// sure the autopilot is not opening channels to itself.
+	Self *btcec.PublicKey
+
+	// PilotCfg is the config of the autopilot agent managed by the
+	// Manager.
+	PilotCfg *Config
+
+	// ChannelState is a function closure that returns the current set of
+	// channels managed by this node.
+	ChannelState func() ([]Channel, error)
+
+	// SubscribeTransactions is used to get a subscription for transactions
+	// relevant to this node's wallet.
+	SubscribeTransactions func() (lnwallet.TransactionSubscription, error)
+
+	// SubscribeTopology is used to get a subscription for topology changes
+	// on the network.
+	SubscribeTopology func() (*routing.TopologyClient, error)
+}
+
+// Manager is struct that manages an autopilot agent, making it possible to
+// enable and disable it at will, and hand it relevant external information.
+// It implements the autopilot grpc service, which is used to get data about
+// the running autopilot, and give it relevant information.
+type Manager struct {
+	started uint32 // To be used atomically.
+	stopped uint32 // To be used atomically.
+
+	cfg *ManagerCfg
+
+	// pilot is the current autopilot agent. It will be nil if the agent is
+	// disabled.
+	pilot *Agent
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+	sync.Mutex
+}
+
+// NewManager creates a new instance of the Manager from the passed config.
+func NewManager(cfg *ManagerCfg) (*Manager, error) {
+	return &Manager{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}, nil
+}
+
+// Start starts the Manager.
+func (r *Manager) Start() error {
+	if !atomic.CompareAndSwapUint32(&r.started, 0, 1) {
+		return nil
+	}
+
+	return nil
+}
+
+// Stop stops the Manager. If an autopilot agent is active, it will also be
+// stopped.
+func (r *Manager) Stop() error {
+	if !atomic.CompareAndSwapUint32(&r.stopped, 0, 1) {
+		return nil
+	}
+
+	if err := r.StopAgent(); err != nil {
+		log.Errorf("Unable to stop pilot: %v", err)
+	}
+
+	close(r.quit)
+	r.wg.Wait()
+
+	return nil
+}
+
+// GetStatus returns the current active status of the autopilot agent.
+func (r *Manager) GetStatus() bool {
```

Comment:
Returns a `bool` atm, instead should possibly be named: `IsActive`?

Reply:
Done!